design/003-interactive-session/003-interactive-session.md

```diff
@@ -0,0 +1,115 @@
+
+The motivation behind this RFC is to provide a simple renku CLI sub-command that basically allows the
+starting and stopping of these interactive sessions.
+
+## Design Detail
+
+The proposed sub-commands of `renku session`, namely to run interactive sessions. The local session heavily relies on
+[docker](https://www.docker.com/). Docker is one of the most popular container engines, that eases the development
+and deployment of applications, by off-loading the installation of the runtime environment dependencies of the
+application.
+
+In order to start a local interactive session, just as in case of renkulab.io, it is required to provide
+a `Dockerfile`. This is by default provided by `renku init` when creating a renku project, but if the user
+is using a custom project template it might not be present.
+
+### Proposed Commands
+
+#### renku session start
+
+A command to start an interactive session.
+
+By default, the `Dockerfile` present in the given project is going to be used for building the docker image
+that's going to be used for running the interactive session. In case the docker image is not available the
+user is going to be prompted whether the user wants to build the docker image, as in some cases it can be
+time consuming.
+
+By using the `--remote [endpoint]` flag an interactive session will be spawned on the specified endpoint. It requires
+that the user has authenticated herself previously with the `renku login` command and aquired the authentication
+token for the endpoint.	If the endpoint is not defined the command will try to assume the endpoint set in the project's
+configuration if available.
+
+##### Detailed Parameter Description
+
+```
+renku session start [OPTIONS]
+
+--image <image_name>	override the docker image to be used for the interactive session.
+--remote [endpoint]	start a remote interactive session using the JWT aquired for the
+			specific endpoint with renku login.
+--cpu <num of cpus>	Specify how much of the available CPU resources a container can use.
+--disk <disk space>	Specify the amount of disk space for the container.
+--gpu <device>		Specify the GPU allocated for the container.
+--memory <max memory>	The maximum amount of memory the container can use.
+
+```
+
+#### renku session stop
+
+`renku session stop my_session` simply stops the given interactive session.
+
+One can shut down all the currently running interactive sessions at once with the `renku session stop --all`.
+
+##### Detailed Parameter Description
+
+```
+renku session stop [OPTIONS] <name>
+
+<name>      Name of the session that the user wants to stop.
+
+--all       Stops all the running containers.
+
+```
+
+#### renku session list
+
+The command to list all the currently running local interactive sessions, with the detailed information regarding
+the network port mappings.
+
+##### Detailed Parameter Description
+
+```
+renku session list [OPTIONS]
+
+--remote       List the remote interactive sessions currently running.
```

Comment:
do you have to specify the remote endpoint here or is it assumed/guessed? Maybe it can be an optional argument?

Reply:
the idea here is that by default only local sessions are listed. for remote session listing the user should explicitly define the `--remote` flag (this is an optional flag)

Reply:
Yes but in `session start` the user specifies a remote

Reply:
Maybe `--remote` should be a flag on `session` and apply to all subcommands

Reply:
> Maybe `--remote` should be a flag on `session` and apply to all subcommands

yeah this should work, although it's a bit 'redudant'/useless in case of `renku session stop <id>` as the <ID> is a unique key for a running session, regardless whether it's remote or local

Reply:
oh I see - where are the session ids stored?

Reply:
These IDs are actually provided by the backend itself. so for example now in case of the local interactive session `renku session list` lists all the running docker containers that are using the image provided by the project itself, i.e. the `Dockerfile`. as stated above these docker images are named the following way: `renku_client.remote['owner']:renku_client.remote['name']:commit_sha` (e.g. `vigsterkr/renku-workflow-demo:9f3d367`).

Note in the meanwhile we've added support to override the default docker image to be used (the one generated based on the project's `Dockerfile`) when starting a session (`renku session start --image renku/renkulab-py:3.9-0ae1458`). This will break the above logic - i.e. listing all the sessions started from the currently active project. But, in case of docker this can be easily fixed by using `--label` when starting a container/session, simply use a the project specific label when creating a container. - will fix this now ASAP.

in case of a remote session this should be fairly easy - as based on the swagger of the notebook api - one can simply query the running servers based on `namespace`, `project` etc.

Reply:
in other words currently I think there's no need to actually store these running session IDs ourselves.

Reply:
> in case of a remote session this should be fairly easy

yes, definitely. But then you need to know which remote to look at. 

Also with regards to local sessions, should we add some more reliable metadata to the container? Or name it in a way that it's identifiable beyond just using the image as a proxy?

Reply:
i've added a simple label in order to support custom images for session:
```
"renku_project": renku_client.remote['owner']:renku_client.remote['name']
```

so for example `"renku_project": "vigsterkr/renku-workflow-demo"`